staging/src/k8s.io/client-go/tools/record/v1beta1/eventseries.go

```diff
@@ -0,0 +1,411 @@
+	tries := 0
+
+	for {
+		if recordEvent(sink, event, b) {
+			break
+		}
+		tries++
+		if tries >= maxTriesPerEvent {
+			klog.Errorf("Unable to write event '%#v' (retry limit exceeded!)", event)
+			break
+		}
+		// Randomize the first sleep so that various clients won't all be
+		// synced up if the master goes down.
+		if tries == 1 {
+			time.Sleep(time.Duration(float64(sleepDuration) * randGen.Float64()))
+		} else {
+			time.Sleep(sleepDuration)
+		}
+	}
+
+}
+
+func recordEvent(sink EventSink, event *v1beta1.Event, isEventSerie bool) bool {
+	var newEvent *v1beta1.Event
+	var err error
+	if isEventSerie {
+
+		patch, err := createPatchBytes(event)
+		if err != nil {
+			klog.Errorf("Unable to calculate diff, no merge is possible: %v", err)
+			return false
+		}
+		newEvent, err = sink.Patch(event, patch)
+	}
+	// Update can fail because the event may have been removed and it no longer exists.
+	if !isEventSerie || (isEventSerie && util.IsKeyNotFoundError(err)) {
+		// Making sure that ResourceVersion is empty on creation
+		event.ResourceVersion = ""
+		newEvent, err = sink.Create(event)
+	}
+	if err == nil {
+
+		event = newEvent
+		return true
+	}
+	panic(err)
+	// If we can't contact the server, then hold everything while we keep trying.
+	// Otherwise, something about the event is malformed and we should abandon it.
+	switch err.(type) {
+	case *restclient.RequestConstructionError:
+		// We will construct the request the same next time, so don't keep trying.
+		klog.Errorf("Unable to construct event '%#v': '%v' (will not retry!)", event, err)
+		return true
+	case *errors.StatusError:
+		if errors.IsAlreadyExists(err) {
+			klog.V(5).Infof("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		} else {
+			klog.Errorf("Server rejected event '%#v': '%v' (will not retry!)", event, err)
+		}
+		return true
+	case *errors.UnexpectedObjectError:
+		// We don't expect this; it implies the server's response didn't match a
+		// known pattern. Go ahead and retry.
+	default:
+		// This case includes actual http transport errors. Go ahead and retry.
+	}
+	klog.Errorf("Unable to write event: '%v' (may retry after sleeping)", err)
+	return false
+
+}
+
+func createPatchBytes(event *v1beta1.Event) ([]byte, error) {
+
+	oldEvent := event.DeepCopy()
+	oldEvent.Series = nil
+
+	oldData, err := json.Marshal(oldEvent)
+	if err != nil {
+		return nil, err
+	}
+	newData, err := json.Marshal(event)
+	if err != nil {
+		return nil, err
+	}
+
+	return strategicpatch.CreateTwoWayMergePatch(oldData, newData, v1beta1.Event{})
+
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (eventBroadcaster *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	eventBroadcaster.fetchEventSeriesForBroadcaster(reportingInstance)
```

Comment:
First of all, I'm not a fan of listing events on every start of component. Listing is very expensive for etcd+apiserver and there can be hundreds of thousands of events (each of them need to be filtered in apiserver). I've already seen cases where listing events was taking minutes if not longer. blocking recorder creation on that seems like a bad decision.

That said, the only alternative is orphaning the existing ongoing series, but there are two consequences:
- the series won't be explicitly stopped (but it will expire in etcd via TTL after some time, so that may not be that huge problem)
- we would create a second series for the same stuff if the condition generating events is still happening (and that is much worse, I don't think it's easily acceptable)

The ideal situation would be an ability to list only "currently ongoing series" without fetching all events.
If at the storage layer we would add a condition that:
- if event is starting a series, we additionally store a separate "series" object in etcd
- if events finishes a series, we remove that corresponding object
That would significantly reduce my fear.
[We would end-up with something kind-of similar to what we've done for services, where we have custom storage:
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/storage/rest.go#L154

@gmarek - when you were writing original proposal, did you have other thoughts about this?

Reply:
This is currently what is done inside `fetchEventSeriesForBroadcaster()` to re-list events
 ```
eventList, err := eventBroadcaster.sink.List(metav1.ListOptions{FieldSelector: fields.Set{"reportingInstance": reportingInstance, "series.state": string(v1beta1.EventSeriesStateOngoing)}.AsSelector().String()})
```

Reply:
Does something like this address your concern @wojtek-t ?

Reply:
Not really - in the end, apiserver has to list everything from etcd and there can be millions of events (and only on those does the filtering).

I need to think more what would be the exact consequences if we would orphan the series - that I think would be much safer from scalability POV.

Reply:
leaving events until the TTL on the key expires, would leave on going series even if in fact they are not, there's a high risk that we mislead users. Can't we rely on etcd to filter on the server-side ?

Reply:
I like @wojtek-t 's idea above of keeping another object for unclosed series and feel it cleanly solves the problem. However, IIUC it seems to have couple of drawbacks:

- it doubles no. of eventseries objects
- it creates 2 extra API calls per event series (to create and delete that "series" object) - which may not turn out in our favour if event series isn't long enough to offset those extra calls

One idea I have to solve above problems is:
- Prefix names of all eventseries coming from a particular source with a unique ID corresponding to that source (for e.g hash of `ReportingInstance` field)
- When the instance restarts (I'm assuming it gets the same ID again), it will need to list all its eventseries. Instead of the expensive listing with filtering, it will list all the relevant objects directly by doing a range query (as we know the key range). But IIUC that will need API changes to carefully allow range queries from client-side (for events).

Would this be an anti-pattern for api? Or is it a feasible option?

Reply:
@shyamjvs for the first solution, when you are verbose serie and the eventrecorder get restarted often, you would create an Event object at each restart.

as for the second solution, we still need to craft a deterministic ID (especially that `reportingInstance` changes depending on where you start the eventrecorder)

cc @bgrant0607

Reply:
Regarding the idea of series: 

This is the design doc:
https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/events-redesign.md
https://github.com/kubernetes/community/pull/932/files

The discussion was here:
https://docs.google.com/document/d/13BeJlrEcJhSKgsHOHWmHdJGqXrjSm_o9XxOtwcN6yNg/edit

Wow, that was a year and a half ago. Trying to page it back in.

If we're trying to reduce the write rate to etcd (as opposed to just the number of unique entities), there's an inherent tradeoff between number of writes and latency and reliability of event reporting. The easiest way to reduce the write rate is to buffer events in the client and flush them lazily once they settle. However, if the controller crashed in the meantime those buffered events would be lost. It would also delay notification of users of problems. So the solution described in the design is a tradeoff, as was the previous (current) deduplication approach.

The first time an event occurs, it is written to etcd. If it recurs, then the idea is to create a series, which is deduplicated in the client, and then written back later.

The design states that "Events with the same <Regarding, Action, Reason, ReportingController, ReportingInstance, Related> tuples" will be treated as part of the same ongoing issue provided their occurrences have sufficient temporal locality. The client can keep track of these in a hash map in memory pretty easily, and occasionally write back the current status. It can keep track of the resource name it used for each event, also. It would only need to do a GET in the case of a resourceVersion conflict or already exists error.
 
I think the main issue is how to mark an event series as completed in the case that the controller crashes before flushing the event series. I don't think that starting a new series when a previous one existed is a significant problem.

One could question whether the distinction between an Ongoing event and a Finished one is even something we should try to record. If there are no additional events, then it's Finished. :-) Or the component died and can't resume reporting the issue. We could deprecate the state field. The API is still beta. 

However, let's say we did want to update it. Use a hash of the key values as part of the object name. No list required. No filtering. Just a single GET. That would occasionally break, such as due to hash function changes (though we could include the function in the name and leave in code to check the old function for a release), but hashing a defined set of fields should be much more forward compatible than hashing the whole object, which is known to not work. If we didn't want to do a GET, we could also just PATCH or DELETE the existing instance.

If we didn't want to do another operation in the event-reporting path, an asynchronous controller could look for incomplete series in the background. It could use the chunking mechanism to retrieve the list in small bits. The design suggests making the events filterable using the series keys. If we ever built indexing that could be made more efficient.

However, if we continue to delete events using the etcd TTL, there would still be cases where some events were in the Ongoing state at deletion time.

Reply:
@bgrant0607 

> One could question whether the distinction between an Ongoing event and a Finished one is even something we should try to record. If there are no additional events, then it's Finished. 

The rationale behind introducing it was that if the series is ongoing, we will be reporting an update once per 30 minutes (by default). Ans if we know it finished, we can just finish it now, instead of waiting for another TTL (1 hour) for it to disappear.
I think it's useful, but I get your point too (I don't have very strong opinion here).

> However, let's say we did want to update it. Use a hash of the key values as part of the object name.

I don't understand how it helps us (well - I think we should do that anyway, but it doesn't solve this particular problem). The reason is that after components restart we don't know what ongoing series we have (that we are currently ongoing).
So while a single GET is enough to extend it (if it's still ongoing), we have no way to get all the ongoing ones that should be finished (because we don't know the items for which they were created). That's my main problem that I see currently.
Does that sound more clear now?

If we would agree that there is no need to report "Finished" state for a series, then that would solve all my concerns.

Reply:
> we have no way to get all the ongoing ones that should be finished

@wojtek-t For this, wdyt about the idea I proposed above?

Reply:
This completely fell off my radar. I think the discussion moved on. The original idea was to list all events matching given set of predicates on fields, which would work if we had a apiserver cache enabled for Events, which we don't. The easiest way to solve this issue is to just enable it, but this will increase the memory usage of the apiserver possibly substantially.

The estimate for memory usage would be: 
100k events (no reason why 100k - it just looks like a reasonable number for a large cluster) * 1kB object size (give or take) = 100MB on a big machine. That's not terrible. For small clusters we probably can say that 10k events is a reasonable number, which will get 10MB memory footprint increase, and that's certainly something that we can live with. I may be off with estimates, but I don't think that I'm off by an order of magnitude.

Dropping the Finished state is not a bad idea, we may just implement client side noting old series as finished for better UX, but that's not strictly necessary.

@shyamjvs idea is basically using our knowledge of the underlying storage implementation to do efficient queries. It may or may not be a good idea, depending on how attached we are to the implementation details of etcd.

Reply:
@shyamjvs - I'm not a huge fan of your suggestion. In order to implement that we will have to leak a bit of etcd-implementation details to a bit higher layes of k8s (i.e. registry). And this doesn't sound like a perfect solution to me.

@gmarek 
re switching on watch-cache of events in memory - please keep in mind that we not only keep state, but also some history log (and we precompute some additional data etc.) Even if we make the history quite small, I think it's still more data that you estimated.
That said, I agree that the numbers doesn't look terrible. Though, what I'm a bit afraid is that we've seen cases where there were waaaay more events in quite small clusters (due to e.g. etcd not removing them as we expected). And I'm afraid that in such cases, apiserver would start oom-ing and the cluster may get completely unusable and harder to recover. I'm not saying "no way", I'm saying we need to be really careful here.

Re dropping finished state - the more I think about it the more I think it's a good idea.
@gmarek how would you do "client side noting old series as finished"? Do you have some ideas about it?, because I actually quite like the idea.

What I just started thinking about is that if we would change components to emit event on start (technically probably on when it is becoming a leader if leader-elected), then any event series whose last event was emitted before the start event of that component may be (client-side) marked as Finished, and all others should already be ongoing.

Thoughts?

Reply:
So you're saying timeout-based - yeah that would be even simpler. That may be enough at least as an initial approach.

> And yes, I saw small clusters with millions of Events because of garbage collection problem, but apiservers were already dead, so I'm not sure if killing them slightly earlier would have been a big deal.

I've seen small clusters with hundreds of thousands of events (in our tests) after https://github.com/kubernetes/kubernetes/pull/64539 was merged where apiserver was still alive. I'm afraid that cases like these may be problematic...
[and yes - I'm not particularly afraid of large clusters that has large masters, those small ones are what concerns me]

Reply:
After a discussing this with @wojtek-t we might want to get numbers on enabling watch cache for events. As for memory footprint, we might want to cache only Event series with an on going state, this should reduce memory usage of the apiserver @gmarek